# **GRIP: The Spark Foundation**
# Author : CHHAVI KUMARI

## **Data Science And Business Analytics Intern**
## TASK 3: EXPLANORATORY DATA ANALYSIS : Retail

In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from scipy.stats import chi2_contingency,chisquare,f_oneway

In [ ]:
pd.set_option('display.max_columns', None)
data=pd.read_csv('SampleSuperstore.csv')
data.head()

Ship Mode    Segment        Country             City       State  \
0    Second Class   Consumer  United States        Henderson    Kentucky   
1    Second Class   Consumer  United States        Henderson    Kentucky   
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   

   Postal Code Region         Category Sub-Category     Sales  Quantity  \
0        42420  South        Furniture    Bookcases  261.9600         2   
1        42420  South        Furniture       Chairs  731.9400         3   
2        90036   West  Office Supplies       Labels   14.6200         2   
3        33311  South        Furniture       Tables  957.5775         5   
4        33311  South  Office Supplies      Storage   22.3680         2   

   Discount    Profit  
0      0.00   41.9136  
1      0.00  219.5820  
2      0.00    6.8714  
3      0.45 -383.0310  
4      0.20    2.5164

In [ ]:
print('Total number of rows:',data.shape[0], 'and columns:', data.shape[1])

Total number of rows: 9994 and columns: 13


In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Ship Mode     9994 non-null   object 
 1   Segment       9994 non-null   object 
 2   Country       9994 non-null   object 
 3   City          9994 non-null   object 
 4   State         9994 non-null   object 
 5   Postal Code   9994 non-null   int64  
 6   Region        9994 non-null   object 
 7   Category      9994 non-null   object 
 8   Sub-Category  9994 non-null   object 
 9   Sales         9994 non-null   float64
 10  Quantity      9994 non-null   int64  
 11  Discount      9994 non-null   float64
 12  Profit        9994 non-null   float64
dtypes: float64(3), int64(2), object(8)
memory usage: 1015.1+ KB


There are 5 Numerical Variables and 8 Categorical Variables.

In [ ]:
data.describe(include = ('int64' ,'float64')) 

Postal Code         Sales     Quantity     Discount       Profit
count   9994.000000   9994.000000  9994.000000  9994.000000  9994.000000
mean   55190.379428    229.858001     3.789574     0.156203    28.656896
std    32063.693350    623.245101     2.225110     0.206452   234.260108
min     1040.000000      0.444000     1.000000     0.000000 -6599.978000
25%    23223.000000     17.280000     2.000000     0.000000     1.728750
50%    56430.500000     54.490000     3.000000     0.200000     8.666500
75%    90008.000000    209.940000     5.000000     0.200000    29.364000
max    99301.000000  22638.480000    14.000000     0.800000  8399.976000

In [ ]:
data.describe(include='object') 

Ship Mode   Segment        Country           City       State  \
count             9994      9994           9994           9994        9994   
unique               4         3              1            531          49   
top     Standard Class  Consumer  United States  New York City  California   
freq              5968      5191           9994            915        2001   

       Region         Category Sub-Category  
count    9994             9994         9994  
unique      4                3           17  
top      West  Office Supplies      Binders  
freq     3203             6026         1523

In [ ]:
data[Sales','Quantity','Discount','Profit'].describe()

Sales     Quantity     Discount       Profit
count   9994.000000  9994.000000  9994.000000  9994.000000
mean     229.858001     3.789574     0.156203    28.656896
std      623.245101     2.225110     0.206452   234.260108
min        0.444000     1.000000     0.000000 -6599.978000
25%       17.280000     2.000000     0.000000     1.728750
50%       54.490000     3.000000     0.200000     8.666500
75%      209.940000     5.000000     0.200000    29.364000
max    22638.480000    14.000000     0.800000  8399.976000

In [ ]:
pd.DataFrame({'Count':data.isnull().sum()}).T

Ship Mode  Segment  Country  City  State  Postal Code  Region  \
Count          0        0        0     0      0            0       0   

       Category  Sub-Category  Sales  Quantity  Discount  Profit  
Count         0             0      0         0         0       0

# 1. Sales vs State

In [ ]:
a=pd.pivot_table(data=data,index=['State'],values=['Sales'],aggfunc='sum')
a=a.sort_values(by='Sales', ascending=False).T
a

State   California    New York        Texas  Washington  Pennsylvania  \
Sales  457687.6315  310876.271  170188.0458   138641.27    116511.914   

State    Florida   Illinois       Ohio   Michigan  Virginia  North Carolina  \
Sales  89473.708  80166.101  78258.136  76269.614  70636.72       55603.164   

State   Indiana   Georgia  Kentucky  New Jersey    Arizona  Wisconsin  \
Sales  53555.36  49095.84  36591.75   35764.312  35282.001   32114.61   

State   Colorado  Tennessee  Minnesota  Massachusetts   Delaware   Maryland  \
Sales  32108.118  30661.873   29863.15      28634.434  27451.069  23705.523   

State  Rhode Island  Missouri  Oklahoma   Alabama    Oregon     Nevada  \
Sales     22627.956  22205.15  19683.39  19510.64  17431.15  16729.102   

State  Connecticut  Arkansas       Utah  Mississippi  Louisiana  Vermont  \
Sales    13384.357  11678.13  11220.056     10771.34    9217.03  8929.37   

State  South Carolina  Nebraska  New Hampshire   Montana  New Mexico     Iowa  \
Sales         8481.71   7464.93       7292.524  5589.352    4783.522  4579.76   

State     Idaho   Kansas  District of Columbia   Wyoming  South Dakota  \
Sales  4382.486  2914.31               2865.02  1603.136       1315.56   

State    Maine  West Virginia  North Dakota  
Sales  1270.53       1209.824        919.91

In [ ]:
print("State having maximum sales = ", a.columns[0])
print("State having minimum sales = ", a.columns[-1])

State having maximum sales =  California
State having minimum sales =  North Dakota


# 2. Profit vs State

In [ ]:
b=pd.pivot_table(data=data,index=['State'],values=['Profit'],aggfunc='sum')
b=b.sort_values(by='Profit', ascending=False).T
b

State   California    New York  Washington    Michigan    Virginia  \
Profit  76381.3871  74038.5486  33402.6517  24463.1876  18597.9504   

State      Indiana     Georgia    Kentucky   Minnesota   Delaware  New Jersey  \
Profit  18382.9363  16250.0433  11199.6966  10823.1874  9977.3748   9772.9138   

State   Wisconsin  Rhode Island   Maryland  Massachusetts   Missouri  \
Profit  8401.8004     7285.6293  7031.1788      6785.5016  6436.2105   

State     Alabama  Oklahoma   Arkansas  Connecticut     Nevada  Mississippi  \
Profit  5786.8253  4853.956  4008.6871    3511.4918  3316.7659    3172.9762   

State        Utah    Vermont  Louisiana   Nebraska    Montana  South Carolina  \
Profit  2546.5335  2244.9783  2196.1023  2037.0942  1833.3285       1769.0566   

State   New Hampshire       Iowa  New Mexico  District of Columbia    Kansas  \
Profit      1706.5028  1183.8119   1157.1161             1059.5893  836.4435   

State      Idaho     Maine  South Dakota  North Dakota  West Virginia  \
Profit  826.7231  454.4862      394.8283      230.1497       185.9216   

State   Wyoming     Oregon    Florida    Arizona  Tennessee   Colorado  \
Profit  100.196 -1190.4705 -3399.3017 -3427.9246 -5341.6936 -6527.8579   

State   North Carolina   Illinois  Pennsylvania        Ohio       Texas  
Profit      -7490.9122 -12607.887   -15559.9603 -16971.3766 -25729.3563

In [ ]:
print("State having maximum profit = ", b.columns[0])
print("State having minimum profit = ", b.columns[-1])

State having maximum profit =  California
State having minimum profit =  Texas


# 3. Quantity vs State

In [ ]:
c=pd.pivot_table(data=data,index=['State'],values=['Quantity'],aggfunc='sum')
c=c.sort_values(by='Quantity', ascending=False).T
c

State     California  New York  Texas  Pennsylvania  Washington  Illinois  \
Quantity        7667      4224   3724          2153        1883      1845   

State     Ohio  Florida  North Carolina  Michigan  Virginia  Arizona  Georgia  \
Quantity  1759     1379             983       946       893      862      705   

State     Colorado  Tennessee  Indiana  Kentucky  Oregon  Massachusetts  \
Quantity       693        681      578       523     499            491   

State     Wisconsin  New Jersey  Maryland  Delaware  Minnesota  Connecticut  \
Quantity        463         454       420       367        331          281   

State     Alabama  Missouri  Oklahoma  Arkansas  Mississippi  Utah  \
Quantity      256       252       247       240          221   219   

State     Rhode Island  South Carolina  Nevada  Louisiana  New Mexico  \
Quantity           199             172     168        156         151   

State     Nebraska  New Hampshire  Iowa  Kansas  Idaho  Montana  Vermont  \
Quantity       136            127   112      74     64       56       50   

State     South Dakota  District of Columbia  Maine  North Dakota  \
Quantity            42                    40     35            30   

State     West Virginia  Wyoming  
Quantity             18        4

In [ ]:
print("State having high quantity sales = ", c.columns[0])
print("State having low quantity sales = ", c.columns[-1])

State having high quantity sales =  California
State having low quantity sales =  Wyoming


# 4. Discount vs State

In [ ]:
d= pd.pivot_table(data=data,index=['State'],values=['Discount'],aggfunc='mean')
d=d.sort_values(by='Discount', ascending=False).T
d

State     Illinois     Texas  Pennsylvania      Ohio  Colorado   Arizona  \
Discount  0.390041  0.370193       0.32862  0.324947  0.316484  0.303571   

State      Florida  Tennessee   Oregon  North Carolina  Wyoming     Idaho  \
Discount  0.299347   0.291257  0.28871        0.283534      0.2  0.085714   

State     West Virginia  California   Montana  Washington    Nevada      Utah  \
Discount          0.075    0.072764  0.066667    0.064032  0.061538  0.060377   

State     New Mexico  New York  Rhode Island  Massachusetts  New Hampshire  \
Discount    0.059459  0.055319      0.021429       0.015556       0.011111   

State     Connecticut  Michigan  Delaware  Maryland  New Jersey  Arkansas  \
Discount     0.007317  0.007059   0.00625  0.005714    0.004615       0.0   

State     Virginia  Vermont  District of Columbia  Wisconsin  South Dakota  \
Discount       0.0      0.0                   0.0        0.0           0.0   

State     South Carolina  Kentucky  North Dakota  Georgia  Oklahoma  Indiana  \
Discount             0.0       0.0           0.0      0.0       0.0      0.0   

State     Louisiana  Iowa  Nebraska  Kansas  Missouri  Mississippi  Minnesota  \
Discount        0.0   0.0       0.0     0.0       0.0          0.0        0.0   

State     Maine  Alabama  
Discount    0.0      0.0

In [ ]:
print("State having maximum average discount = ", d.columns[0])
print("State having minimum average discount = ", d.columns[-1])

State having maximum average discount =  Illinois
State having minimum average discount =  Alabama


## Key Inferences from 1, 2, 3 & 4:

1. California is having highest profit among 59 states, the reason behind is high quantity sales in that state.
2. Despite giving highest average discount in Illinois, the state is not having enough sales and profit. This need to be review by marketing teams.
3. Wyoming is having low quantity sales as well as low total sales. 
4. Similarly North Dakota is having lowest total sales. So commercial advertising and giving more discount can lift up the sales in these region.
5. Texas is suffering from huge loss. May be giving more discount or spending more on shipment mode are responsible. Further it has 3rd highest total sales among 59 states.


# 5. Sub-Category vs Profit

In [ ]:
e= pd.pivot_table(data=data,index=['Sub-Category'],values=['Profit'],aggfunc='sum')
e=e.sort_values(by='Profit', ascending=False).T
e

Sub-Category     Copiers      Phones  Accessories       Paper     Binders  \
Profit        55617.8249  44515.7306   41936.6357  34053.5693  30221.7633   

Sub-Category      Chairs     Storage  Appliances  Furnishings  Envelopes  \
Profit        26590.1663  21278.8264  18138.0054   13059.1436  6964.1767   

Sub-Category       Art    Labels   Machines  Fasteners   Supplies  Bookcases  \
Profit        6527.787  5546.254  3384.7569   949.5182 -1189.0995  -3472.556   

Sub-Category      Tables  
Profit       -17725.4811

In [ ]:
print("Sub-Category which makes maximum profit = ", e.columns[0])
print("Sub-Category which makes minimum profit = ", e.columns[-1])

Sub-Category which makes maximum profit =  Copiers
Sub-Category which makes minimum profit =  Tables


# 6. Sub-Category vs Discount

In [ ]:
f= pd.pivot_table(data=data,index=['Sub-Category'],values=['Discount'],aggfunc='mean')
f= f.sort_values(by='Discount', ascending=False).T
f

Sub-Category   Binders  Machines    Tables  Bookcases    Chairs  Appliances  \
Discount      0.372292  0.306087  0.261285    0.21114  0.170178    0.166524   

Sub-Category   Copiers    Phones  Furnishings  Fasteners  Envelopes  \
Discount      0.161765  0.154556     0.138349   0.082028   0.080315   

Sub-Category  Accessories  Supplies     Paper       Art   Storage    Labels  
Discount         0.078452  0.076842  0.074891  0.074874  0.074704  0.068681

In [ ]:
print("Sub-category having maximum average discount = ", f.columns[0])
print("Sub-category having minimum average discount = ", f.columns[-1])

Sub-category having maximum average discount =  Binders
Sub-category having minimum average discount =  Labels


# 7. Segment vs Profit

In [ ]:
g=pd.pivot_table(data=data,index=['Segment'],values=['Profit'],aggfunc='sum')
g=g.sort_values(by='Profit', ascending=False).T
g

Segment     Consumer  Corporate  Home Office
Profit   134119.2092  91979.134   60298.6785

In [ ]:
print("Segment which makes maximum profit = ", g.columns[0])
print("Segment which makes minimum profit = ", g.columns[-1])

Segment which makes maximum profit =  Consumer
Segment which makes minimum profit =  Home Office


## Key Inferences from 5, 6 & 7:
1. Copiers is high profit making sub-category. While selling tables is giving us a huge loss.
2. Company is giving high discount rate on Tables (3rd highest). That may be the reason behind this much loss.
3. Consumer segments are giving the higher profit. While home office is giving half profit as cosumer segment.
4. Company need to take care of Home office segment and target their costumers by making good quality products or attracting them by advertisments.


In [ ]:
# Now building another column "Overall" for better analysis
# Overall column will be categorical column showing gain or loss depending on profit values.
data['overall']=data['Profit'].apply(lambda x: 'gain' if x>0 else 'loss')

In [ ]:
data.head(10)

Ship Mode    Segment        Country             City       State  \
0    Second Class   Consumer  United States        Henderson    Kentucky   
1    Second Class   Consumer  United States        Henderson    Kentucky   
2    Second Class  Corporate  United States      Los Angeles  California   
3  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
4  Standard Class   Consumer  United States  Fort Lauderdale     Florida   
5  Standard Class   Consumer  United States      Los Angeles  California   
6  Standard Class   Consumer  United States      Los Angeles  California   
7  Standard Class   Consumer  United States      Los Angeles  California   
8  Standard Class   Consumer  United States      Los Angeles  California   
9  Standard Class   Consumer  United States      Los Angeles  California   

   Postal Code Region         Category Sub-Category     Sales  Quantity  \
0        42420  South        Furniture    Bookcases  261.9600         2   
1        42420  South        Furniture       Chairs  731.9400         3   
2        90036   West  Office Supplies       Labels   14.6200         2   
3        33311  South        Furniture       Tables  957.5775         5   
4        33311  South  Office Supplies      Storage   22.3680         2   
5        90032   West        Furniture  Furnishings   48.8600         7   
6        90032   West  Office Supplies          Art    7.2800         4   
7        90032   West       Technology       Phones  907.1520         6   
8        90032   West  Office Supplies      Binders   18.5040         3   
9        90032   West  Office Supplies   Appliances  114.9000         5   

   Discount    Profit overall  
0      0.00   41.9136    gain  
1      0.00  219.5820    gain  
2      0.00    6.8714    gain  
3      0.45 -383.0310    loss  
4      0.20    2.5164    gain  
5      0.00   14.1694    gain  
6      0.00    1.9656    gain  
7      0.20   90.7152    gain  
8      0.20    5.7825    gain  
9      0.00   34.4700    gain

# 8. Statistical analysis for categorical data types, Chi-square Test is performed.
###  We will perform Chi-Square test to show dependency of other Categorical columns on "Overall" column.

In [ ]:
cat_cols = list(data.describe(include = "O").columns)
print(cat_cols)

['Ship Mode', 'Segment', 'Country', 'City', 'State', 'Region', 'Category', 'Sub-Category', 'overall']


In [ ]:
chi_stat=[]
p_value=[]
for i in cat_cols:
    chi_res=chi2_contingency(np.array(pd.crosstab(data[i],data['overall'])))
    chi_stat.append(chi_res[0])
    p_value.append(chi_res[1])
chi_square=pd.DataFrame([chi_stat,p_value])
chi_square=chi_square.T
col=['Chi Square Value','P-Value']
chi_square.columns=col
chi_square.index=cat_cols

In [ ]:
chi_square

Chi Square Value        P-Value
Ship Mode            17.365927   5.942422e-04
Segment               3.096504   2.126193e-01
Country               0.000000   1.000000e+00
City               2614.630482  6.349617e-272
State              2626.352325   0.000000e+00
Region              416.444920   6.065863e-90
Category            433.036350   9.275800e-95
Sub-Category       1930.724475   0.000000e+00
overall            9987.598577   0.000000e+00

In [ ]:
chi_square[chi_square["P-Value"]<0.05]

Chi Square Value        P-Value
Ship Mode            17.365927   5.942422e-04
City               2614.630482  6.349617e-272
State              2626.352325   0.000000e+00
Region              416.444920   6.065863e-90
Category            433.036350   9.275800e-95
Sub-Category       1930.724475   0.000000e+00
overall            9987.598577   0.000000e+00

In [ ]:
features_p = list(chi_square[chi_square["P-Value"]<0.05].index)
print("Significant categorical Features:\n",features_p)

Significant categorical Features:
 ['Ship Mode', 'City', 'State', 'Region', 'Category', 'Sub-Category', 'overall']


## Key Inference from 8:
1. Categorical feature with P-value<0.05, i.e. {'Ship Mode', 'City', 'State', 'Region', 'Category', 'Sub-Category'} are dependent and have correlation with Overall gain and loss.

# 9. Statistical analysis for numerical data types , ANOVA Test is performed
###  We will perform ANOVA test to show dependency of other Numerical columns on "Overall" column.

In [ ]:
num_cols = list(data.describe(include = ('int64','float64')).columns)
print(num_cols)

['Postal Code', 'Sales', 'Quantity', 'Discount', 'Profit']


In [ ]:
f_stat=[]
p_val=[]
for i in num_cols:
    atr_0=data[data['overall']=="loss"][i]
    atr_1=data[data['overall']=="gain"][i]
    a=f_oneway(atr_0,atr_1)
    f_stat.append(a[0])
    p_val.append(a[1])
anova=pd.DataFrame([f_stat,p_val])
anova=anova.T
cols=['F-STAT','P-VALUE']
anova.columns=cols
anova.index=num_cols

In [ ]:
anova

F-STAT        P-VALUE
Postal Code      1.229693   2.674935e-01
Sales            4.356861   3.688544e-02
Quantity         0.224060   6.359737e-01
Discount     12295.809791   0.000000e+00
Profit         551.540309  9.229395e-119

In [ ]:
anova[anova["P-VALUE"]<0.05]

F-STAT        P-VALUE
Sales         4.356861   3.688544e-02
Discount  12295.809791   0.000000e+00
Profit      551.540309  9.229395e-119

In [ ]:
features_p_n = list(anova[anova["P-VALUE"]<0.05].index)
print("Significant numerical Features:\n",features_p_n)

Significant numerical Features:
 ['Sales', 'Discount', 'Profit']


## Key Inference from 9:
1. Numerical feature with P-value<0.05, i.e. {'Sales', 'Discount', 'Profit'} are dependent and have correlation with Overall gain and loss.


# 10. Graphical Dashboard and Analysis of all features:

In [ ]:
!pip install pywedge

In [ ]:
import pywedge as pw

In [ ]:
dash=pw.Pywedge_Charts(data, c=None, y='Profit')

In [ ]:
charts=dash.make_charts()

HTML(value='<h2>Pywedge Make_Charts </h2>')

## Key Inferences from 10:

1.   **Scatter Chart:**

*   Higher discount is negatively affecting sales and profit.
*   Costumers are not getting attracted by high discount
*   People tends to buy product with smaller discount.


2.   **Pie chart**

*   Consumer segment are making maximum profit with 54.3% of total profit on segments while Corporate segment are making minimum profit with 28.5%.
*   Office supplies are making maximum profit with 38% of total profit on categories while Furniture are making minimum profit with 25.7%.
*   Callifornia is making maximum profit with 14.5% of total profit on states .
*   Phones sub-category are making maximum profit with 14.2% of total profit on sub-categories while Supplies sub-category are making minimum profit with 0.13%.

3.   **Bar plot**

*   West region with Category "Technology" is suffering huge loss.
*   East region with Category "Office supplies" and "Technology" are making good profit.
*   Home office with category "Technology" is suffering huge loss while Consumer with category "Technology" is making good profit.

4.  **Histogram**

*   Binders are maximum selling product.
*   In callifornia,we are selling maximum products while In North Dakota,we are selling minimum products.

5.  **Correlation plot**

*   Profit is negatively correlated with discount means items with high discount are making less profit and vice-versa.
*   Quantity with sales and Profit with sales are positively correlated.




# THE END